File: cpp/cpp_10-99/33.cc
// C++
// leetcode 33
// https://leetcode-cn.com/problems/search-in-rotated-sorted-array/
// 思路:
// 经典二分法
// 不同之处是第一层的二分要做多一点判断

class Solution {
public:
	int bisearch(vector<int>& nums, int low, int high, int target){
		while(low <= high)
		{
			int mid = low + (high - low) / 2;
			if(nums[mid] == target)	return mid;
			else if(nums[mid] < target)	low = mid + 1;
			else	high = mid - 1;
		}
		return -1;
	}

	int inside(vector<int>& nums, int l, int r, int target){
		if(l > r)	return -1;
		int mid = l + (r - l) / 2;
		if(target == nums[mid])	return mid;
		if(target <= nums[r] && target > nums[mid])
			return bisearch(nums, mid + 1, r, target);
		if(target >= nums[l] && target < nums[mid])	
			return bisearch(nums, l, mid - 1, target);
		if(nums[mid] < nums[l])
			return inside(nums, l, mid - 1, target);
		else
			return inside(nums, mid + 1, r, target);
	}

    int search(vector<int>& nums, int target) {
    	return inside(nums, 0, nums.size() - 1, target);
    }
};

File: Go/Go_10-99/33.go
// Go
// leetcode 33
// https://leetcode.cn/problems/search-in-rotated-sorted-array/

func search(nums []int, target int) int {
	n := len(nums)
	idx := search_helper(nums,0,n-1, target)
	if idx >= n || idx < 0 || nums[idx] != target {
		return -1
	}
    return idx
}

func search_helper(arr []int, l, r, t int) int {
    // fmt.Println(l, r)
	if l == r {
		if arr[l] == t {
			return l
		}
		return -1
	}
	if arr[l] < arr[r] {
		return biSearch(arr, l, r, t)
	}
	if arr[l] == t {
		return l
	}
	if arr[r] == t {
		return r
	}

	mid := l + (r-l)/2
	if arr[l] < arr[mid] { // 左 good， 右旋转
		x := biSearch(arr, l ,r, t)
		if x != len(arr) && arr[x] == t {
			return x
		}
		return search_helper(arr, mid, r, t)
	}

	// 左旋转，右 good
	x := biSearch(arr, l, r, t)
	if x != len(arr) && arr[x] == t {
		return x
	}
	return search_helper(arr, l, mid, t)
}


func biSearch(arr []int, l, r, target int) int {
	for l < r {
		mid := l + (r-l)/2
		if arr[mid] < target {
			l = mid + 1
		} else {
			r = mid
		}
	}
	return l
}